src/lib/core/media-marshaller/media-marshaller.ts

```diff
@@ -0,0 +1,175 @@
+    }
+    this.updateStyles();
+  }
+
+  /**
+   *
+   * @param element
+   * @param key
+   * @param builder optional so that custom bp directives don't have to reprovide this
+   * @param watcher
+   */
+  init(element: HTMLElement, key: string, builder?: Function, watcher?: MarshalWatcher): void {
+    if (builder) {
+      let builders = this.builderMap.get(element);
+      if (!builders) {
+        builders = new Map();
+        this.builderMap.set(element, builders);
+      }
+      builders.set(key, builder);
+    }
+    if (watcher && !this.watcherMap.get(element)) {
+      const observables = watcher.observables;
+      if (watcher.parent) {
+        const parentSubject = this.parentSubjects.get(watcher.parent) || new Subject();
+        observables.push(parentSubject.asObservable());
+        this.parentSubjects.set(watcher.parent, parentSubject);
+      }
+      const subscription = merge(...observables).subscribe(() => {
+        const currentValue = this.getValue(element, key);
+        this.updateElement(element, key, currentValue);
+      });
+      this.watcherMap.set(element, subscription);
+    }
+  }
+
+  /**
+   * get the value for an element and key and optionally a given breakpoint
+   * @param element
+   * @param key
+   */
+  getValue(element: HTMLElement, key: string): string {
+    const bpMap = this.elementMap.get(element);
+    if (bpMap) {
+      const values = this.getFallback(bpMap);
+      if (values) {
+        return values.get(key) || '';
+      }
+    }
+    return '';
+  }
+
+  /**
+   * Set the value for an input on a directive
+   * @param element the element in question
+   * @param key the type of the directive (e.g. flex, layout-gap, etc)
+   * @param bp the breakpoint suffix (empty string = default)
+   * @param val the value for the breakpoint
+   */
+  setValue(element: HTMLElement, bp: string, key: string, val: any): void {
+    let bpMap: BreakpointMap | undefined = this.elementMap.get(element);
+    if (!bpMap) {
+      bpMap = new Map().set(bp, new Map().set(key, val));
+      this.elementMap.set(element, bpMap);
+    } else {
+      const values = (bpMap.get(bp) || new Map()).set(key, val);
+      bpMap.set(bp, values);
+      this.elementMap.set(element, bpMap);
+    }
+    this.updateElement(element, key, this.getValue(element, key));
+  }
+
+  updateStyles(): void {
+    this.elementMap.forEach((bpMap, el) => {
+      const valueMap = this.getFallback(bpMap);
+      if (valueMap) {
+        valueMap.forEach((v, k) => this.updateElement(el, k, v));
+      }
+    });
+  }
+
+  updateElement(element: HTMLElement, key: string, val: any): void {
+    const builders = this.builderMap.get(element);
+    if (builders) {
+      const builder = builders.get(key);
+      if (builder) {
+        builder(val);
+        const parentSubject = this.parentSubjects.get(element);
+        if (parentSubject) {
+          parentSubject.next();
+        }
+      }
+    }
+  }
+
+  destroy(): void {
```

Comment:
This needs to be fleshed out due to the subscribe-y nature of this, but also the ephemeral nature of services.

Reply:
We should talk about this one.